design-proposals/klip-0-add-protobuf-support.md

```diff
@@ -0,0 +1,77 @@
+
+## tl;dr
+
+Add support to KSQL to process messages stored using the [ProtocolBuffers](https://developers.google.com/protocol-buffers/) serialization format.   
+
+## Motivation and background
+
+[ProtocolBuffers](https://developers.google.com/protocol-buffers/) is a language and platform neutral serialization format that is commonly used with Apache Kafka. Adding out-of-the-box support for this message format would expand the environments in which KSQL can be utilized.
+
+## What is in scope
+
+Adding SerDe support for ProtocolBuffers Version 3 to KSQL.
+
+## What is not in scope
+
+Supporting streams/topics containing multiple ProtocolBuffer schema types.  _See [Design](#Design) section below for more details._
+
+## Value/Return
+
+Out of the box support for running KSQL against Apache Kafka topics containing messages encoded with ProtocolBuffers.
+
+## Public APIS
+
+Proposed extending the `CREATE STREAM` grammar to include a new _value_format_ type **PROTOBUF** as well as adding a new _tableProperty_ named **protobuf_class**.
+
+Example:
+```genericsql
+ksql> CREATE STREAM stream_name (....) WITH (kafka_topic='TopicName', value_format='PROTOBUF', protobuf_class='org.whatever.protobufs.MyProtoBuf$EncodedMessageType');
+```
+
+## Design  
+
+Before digging into the design, it's important to understand how ProtocolBuffers work. A ProtocolBuffer encoded message does **not** include the schema definition. A consumer attempting to deserialize a ProtocolBuffer encoded message must know in advance the schema of that message. Because ProtocolBuffer messages lack this identifying information, services like SchemaRegistry do not exist for ProtocolBuffers.   
+
+**Important:** Because of the limitations in being unable to identify ProtocolBuffer schemas from an encoded message, the proposed design will be limited to only support processing messages from a stream encoded with a single message type/schema.
+
+When defining a new stream, the user would define the following _tableProperties_:
+- value_format: **PROTOBUF**
+- protobuf_class: fully qualified classname for the ProtocolBuffer class represented in the stream's messages.
+
+
+The class defined by the _protobuf_class_ tableProperty would need to already exist at runtime within KSQL's classpath.
+
+When the Serializer/Deserializer instances are created the _protobuf_class_ property would be passed to the KsqlProtobufSerializer/Deserializer via the `public void configure(final Map<String, ?> properties, ...)` method. The Serializer/Deserializer would then be able to instantiate an instance of the configured ProtocolBuffer class as needed. 
+
+The Deserializer `public GenericRow deserialize(final String topic, final byte[] bytes)` implementation would accept a byte array representing a serialized ProtocolBuffer. This method would deserialize the byte array into the configured ProtocolBuffer instance, and apply the appropriate transformation into its equivalent `GenericRow` representation.
```

Comment:
Out plan is to drop `GenericRow` some time soon. As such, I don't think we'd want to expose `GenericRow` on any interfaces. We'd probably go with a Struct type, but this will need to be discussed and decided upon.

(Same goes for the serialize call, obvs)

Reply:
I think the document would benefit from some additional detail about how a given proto class would be converted to a `GenericRow` or its `Struct` like replacement.

i.e. I'm assuming this would enumerate the fields in the proto class and expose these as columns to KSQL? 

How to the types available in PB map to KSQL types?

Reply:
> How to the types available in PB map to KSQL types?

@Crim Can you give me write access to your fork? I would be willing to flesh that out.

Reply:
Added you as a collaborator on my fork.

Reply:
> Out plan is to drop GenericRow some time soon. As such, I don't think we'd want to expose GenericRow on any interfaces. We'd probably go with a Struct type, but this will need to be discussed and decided upon.

Well the current implementation (and this proposal) are not that special. Basically it implements a `KsqlProtobufTopicSerDe` (similar to something like say `KsqlAvroTopicSerDe`).
The one point where it differs from - say - Avro, is that it directly implements the (De)serializers instead of using the Kafka Connect implementations (probably because Confluent has no Protobuf Connector available).

So yes, there would be new (De)serializers. The interface, which exposes `GenericRow` are basically all interfaces that already exist, though. And you could make the new (De)serializers `private` and not expose more information than currently.

Personally, I think the current impl is great because it is very much straight forward instead of jumping through hoops for functionality that nobody uses. Then again I am not a Java Dev by trade :)